// adaptivejs/tasks/config/connect.js
module.exports = function(grunt){
    var buildRunning = false;
    var maxDelay = 15000;

    grunt.event.on('watch-start', function() { 
        buildRunning = true;
        setTimeout(function() { buildRunning = false; }, maxDelay);
    });

    grunt.event.on('watch-end', function() {
        buildRunning = false;
    });

    var middlewareSetup = function(connect, options, middlewares) {

        var enableCrossDomainRequests = function (req, res, next) {
            // CSOPS-103: Chrome now blocks cross origin requests for resources
            // https://code.google.com/p/chromium/issues/detail?id=286681
            res.setHeader('Access-Control-Allow-Origin', '*');
            return next();
        };

        var buildDelay = function(req, res, next) {
            var retryInterval = 100;
            if(buildRunning){
                setTimeout(function(){ buildDelay(req, res, next); }, retryInterval);
            } else {
                return next();
            }
        };

        middlewares.unshift(enableCrossDomainRequests);
        middlewares.unshift(buildDelay);

        return middlewares;
    };

    return {
        http: {
            options: {
                hostname: '0.0.0.0',
                port: (grunt.option('port') || 8080),
                base: 'build/',
                middleware: middlewareSetup,
                debug: grunt.option('debug') || false
            }
        },
        https: {
            options: {
                hostname: '0.0.0.0',
                protocol: 'https',
                port: (grunt.option('https-port') || 8443),
                base: 'build/',
                middleware: middlewareSetup,
                debug: grunt.option('debug') || false
            }
        },
        test: {
            options: {
                hostname: '0.0.0.0',
                port: (grunt.option('p') || 8888),
                base: '.'
            }
        }
    };
};
