// adaptivejs/tasks/config/requirejs.js

module.exports = function(grunt) {
    var loaderOptions = {
        // almond: true,
        mainConfigFile: './app/config/loader.js',
        baseUrl: 'app/',
        wrap: true,
        keepBuildDir: true,

        optimize: 'none',
        preserveLicenseComments: false,
        generateSourceMaps: true,

        name: 'loader',
        // Called "adaptive.js" to match filename in tag
        out: './build/adaptive.js',
        onModuleBundleComplete: function (data) {
            var fs = require('fs'),
                amdclean = require('amdclean'),
                outputFile = data.path;

            console.log(outputFile);
            fs.writeFileSync(outputFile, amdclean.clean({
              'filePath': outputFile
            }));
        },
        paths: {
            'settings': '../node_modules/adaptivejs/lib/settings',
            'adaptivejs': '../node_modules/adaptivejs/src',
            'mobifyjs/utils': '../node_modules/adaptivejs/node_modules/mobifyjs-utils/utils',
            'preload-html': '../build/preload.min.html',
            'preload-css': '../build/preload.min.css',
            'preload-js': '../build/preload.min.js',
            'text': '../node_modules/adaptivejs/bower_components/requirejs-plugins/lib/text'
        },
        stubModules: ['text']
    };
    var adaptationOptions = {
        almond: true,
        mainConfigFile: './app/config/adaptation.js',
        baseUrl: 'app/',
        wrap: true,
        keepBuildDir: true,

        optimize: 'none',
        preserveLicenseComments: false,
        generateSourceMaps: true,

        // Export Almond
        wrap: {
            start:
                "(function(){\n",
            end:
                "   window.Adaptive = window.Adaptive || {};\n" +
                "   window.Adaptive.AMD = {};\n" +
                "   window.Adaptive.AMD.require = require;\n" +
                "   window.Adaptive.AMD.define = define;\n" +
                "})();"
        },

        name: 'adaptation',
        out: './build/adaptation.js',
        paths: {
            'text': '../node_modules/adaptivejs/bower_components/requirejs-plugins/lib/text',
            'json': '../node_modules/adaptivejs/bower_components/requirejs-plugins/src/json',
            'mobifyjs/utils': '../node_modules/adaptivejs/node_modules/mobifyjs-utils/utils',
            'devSettings': 'devSettings',
            'settings': '../node_modules/adaptivejs/lib/settings',
            'adaptivejs': '../node_modules/adaptivejs/src',
            '$': '../node_modules/adaptivejs/src/selectorLibrary',
            'dust': '../node_modules/adaptivejs/lib/dust-requirejs',
            'dust-custom': '../node_modules/adaptivejs/lib/dustPatch',
            'dust-core': 'bower_components/dustjs-linkedin/dist/dust-core',
            'dust-full': 'bower_components/dustjs-linkedin/dist/dust-full',
        },
        shim: {
            'dust-core': {
                'exports': 'dust'
            },
            'dust-full': {
                'exports': 'dust'
            }
        },
        stubModules: ['text', 'json', 'dust'],
        exclude: ['dust-full']
    };

    var uiOptions = {
        removeCombined: true,
        name: 'ui',
        mainConfigFile: './app/config/ui.js',
        baseUrl: 'app/',

        optimize: 'none',
        preserveLicenseComments: false,
        generateSourceMaps: true,

        // Import Almond
        wrap: {
            start:
                "(function(require, define, window, document) {\n",
            end:
                "})(Adaptive.AMD.require" +
                ", Adaptive.AMD.define" +
                ", window" +
                ", document" +
                ")\n\n"
        },
        out: './build/js/ui.js',
        paths: {
            'external': '../node_modules/adaptivejs/lib/external-requirejs',
            'dust': '../node_modules/adaptivejs/lib/dust-requirejs',
            'dust-full': 'bower_components/dustjs-linkedin/dist/dust-full',
            'dust-custom': 'empty:',
            'dust-core': 'empty:',
            '$': 'empty:'
        },
        shim: {
            'dust-core': {
                'exports': 'dust'
            },
            'dust-full': {
                'exports': 'dust'
            }
        },
        stubModules: ['dust'],
        exclude: ['dust-full']
    };

    return {
        // Building adaptive.js (the loader)
        loader: {
            options: loaderOptions
        },
        loaderMin: {
            options: grunt.util._.extend(
                grunt.util._.clone(loaderOptions),
                {
                    optimize: 'uglify2',
                    generateSourceMaps: false,
                    out: './build/adaptive.min.js'
                }
            )
        },
        // Building adaptation.js
        adaptation: {
            options: adaptationOptions
        },
        adaptationMin: {
            options: grunt.util._.extend(
                grunt.util._.clone(adaptationOptions),
                {
                    optimize: 'uglify2',
                    generateSourceMaps: false,
                    out: './build/adaptation.min.js',
                    paths: grunt.util._.extend(
                        grunt.util._.clone(adaptationOptions.paths),
                        {
                            'settings': '../node_modules/adaptivejs/lib/stubSettings'
                        }
                    )
                }
            )
        },
        // Building front-end UI scripts (ui.js)
        ui: {
            options: uiOptions
        },
        uiMin: {
            options: grunt.util._.extend(
                grunt.util._.clone(uiOptions),
                {
                    optimize: 'uglify2',
                    generateSourceMaps: false,
                    out: './build/js/ui.min.js'
                }
            )
        }

    };
};
